### ISSUE TITLE
Aliasing in Fusion

### ISSUE THREAD

**Issue by jjsjann123** [csarofeen/pytorch#929]
## 🚀 Feature

Our current fusion code supports outputs aliasing to inputs in a bare bone manner. We only support aliasing in the integration code, where output tensor will reuse storage from aliased inputs, instead of allocating fresh tensor. This allows us to mimic in-place update, while exposing codegen to racing issues. Currently it is used in a very limited scope, in-place update to running stats in batch normalization. Since running stats are not used anywhere else within the fusion group, it's safe to do so.

With our recent addition of autocast operation, which has a potential to introduce another form of aliasing, we need to expand its support in integration code.
In the case of a no-op autocast, the operation is supposed to return an alias to the input tensor.

In this made-up example below as a fusion:
```
    %o_0 = fusible_op(...)
    %o_1 = autocast_to_fp32(%o_0)
    return (%o_0, %o_1)
```
While we parse the graph, we should know when %o_0 is in fp32 already, hence the op is a no-op and we should just map %o_1 to the same TensorView of %o_0.
Right now, we are making copies of input to output for cast op. This has a risk of producing inconsistent output comparing to a native PyTorch JIT execution.

## Pitch

I think it's safe to implement the alias support in integration code, although it sounds pretty scary for codegen :). The confusing parts would be 1. the discrepancy of the scalar type of TensorView and the scalar type of JIT Tensor; 2. Fusion output need to be casted back to their JIT scalar type, we need to carefully think about how to update the alias map then.

## Additional context

This topic is started in the PR #879 and have also been brought up in #914. Our current plan is to NOT prioritize this until we find an model where it becomes a real problem.
